### ISSUE TITLE
Problem with zetteldeft.el installation

### ISSUE THREAD

**Issue by Phil2740** [EFLS/zetteldeft#82]
Hi all,
This is a newbie question, I apologise in advance if there is something I should have seen before.

I upgraded Zetteldeft today (2020-10-03). The function zetteldeft-set-classic-keybindings seems to have disappeared. It doesn't show up when I try C-h f.

What have I missed? I'm only just getting to know the classic key bindings. I understand that I could re-bind all the keys manually. How can I get the function back?

Thanks for your help.

Phil

**Comment by EFLS** [comment]
That's strange, I haven't touched it so no clue what might have caused the change.

How did you upgrade (and what was your previous version)? How are you loading Zetteldeft? MELPA?

**Comment by Phil2740** [comment]
Thanks for your reply.

I loaded  Zetteldeft via MELPA and upgraded with the Package options on the Emacs menu bar. I know It was only a couple of days ago, but I seem to remember that in the upgrade process, the avy package got deleted, and I had to re-install that package. I’m sure that when I first loaded Zetteldeft (some months ago, but it lay dormant for a while) all the dependencies were installed.    

The version I have is version 0.3. I don’t remember what the previous version was, perhaps I never looked.

**Comment by EFLS** [comment]
Aight, some other questions. Do you load the package in your `init.el`? Are you sure the package is loaded? I mean, can you use the other functions?

What happens when you hit `M-x zetteldeft`?

**Comment by Phil2740** [comment]
I am sure the package gets loaded from my init.el. Here is the message that I get in the *Warnings* buffer when I first fire up Emacs:

Warning (initialization): An error occurred while loading ‘/Users/xxxxxx/.emacs.d/init.el’:

Symbol's function definition is void: zetteldeft-set-classic-keybindings

To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace.

Other functions seem to work. If I type M-x deft I get a listing of my zetteldeft directory. M-x zetteldeft-tag-buffer gives me the tag buffer. I didn’t try all the available functions.

I use Helm. If I type M-x zetteldeft I get these entries in the *helm M-x* buffer:

zetteldeft-new-file
zetteldeft-find-file
zetteldeft-tag-buffer
zetteldeft-follow-link
zetteldeft-count-words
zetteldeft-file-rename
zetteldeft-org-graph-note
zetteldeft-avy-tag-search
zetteldeft-avy-link-search
zetteldeft-avy-file-search
zetteldeft-deft-new-search
zetteldeft-search-at-point
zetteldeft-org-graph-search
zetteldeft-org-search-insert
zetteldeft-insert-list-links
zetteldeft-new-file-and-link
zetteldeft-search-current-id
zetteldeft-org-search-include
zetteldeft-find-file-id-insert
zetteldeft-copy-id-current-file
zetteldeft-insert-list-links-missing
zetteldeft-avy-file-search-ace-window
zetteldeft-find-file-full-title-insert

There is no sign of the new(ish) function zetteldeft-new-file-and-backlink, which I find really useful.

I had a look at zetteldeft.el, I couldn’t find any function named zetteldeft-set-classic-key-bindings. I suppose this is the file I downloaded from MELPA.

I hope this helps with a diagnosis.

**Comment by EFLS** [comment]
> There is no sign of the new(ish) function zetteldeft-new-file-and-backlink, which I find really useful.

> I had a look at zetteldeft.el, I couldn’t find any function named zetteldeft-set-classic-key-bindings. I suppose this is the file I downloaded from MELPA.

That doesn't sound right. My guess is that the wrong `zetteldeft.el` is being loaded. Both functions should be there (just double checked in the code from MELPA).

Can you share the relevant code from your `init.el`? Also, can you find the file `zetteldeft-pkg.el` and check what it says on the first line? It should mention version info, like: `"20200922.712"`.

**Comment by Phil2740** [comment]
Here is the complete version of zetteldeft-pkg.el that is on my system:

(define-package "zetteldeft" "0.3" "Turn deft into a zettelkasten system"
  '((emacs "25.1")
    (deft "0.8"))
  :commit "271ea573b0a4f265d16108db2ec7c928f3e9aa31" :keywords
  '("deft" "zettelkasten" "zetteldeft" "wp" "files")
  :authors
  '(("EFLS <Elias Storms>"))
  :maintainer
  '("EFLS <Elias Storms>")
  :url "https://efls.github.io/zetteldeft/")
;; Local Variables:
;; no-byte-compile: t
;; End:

I don’t know whether it has any bearing on the problem, but I am running Emacs 27.1, so I am unsure what effect the command emacs (“25.1”) would have.

And here is the section of my init.el to do with zetteldeft:

;; zetteldeft
;; provides a way to create and manage links between short notes
;; source: eliasstorms.net
;; added 2 May 2020

(require 'deft)
(setq deft-directory "~/Dropbox/zettel")

;; this section from Pragmatic Emacs: Deft as a File Search Tool
;; added 28 May 2020
(setq deft-extensions '("org"))
(setq deft-default-extension "org")
(setq deft-text-mode 'org-mode)
(setq deft-use-filename-as-title t)
(setq deft-use-filter-string-for-filename t)
(setq deft-auto-save-interval 0)

;; set deft sort method
;; source: deft documentation
;; added 25 September 2020
;; (setq deft-current-sort-method 'title)

(require 'zetteldeft)
(zetteldeft-set-classic-keybindings)

;; more zetteldeft key bindings
;; added by PC, 25 September 2020
(global-set-key (kbd "C-c d b") 'zetteldeft-new-file-and-backlink)

My knowledge of Elisp is very limited, this code may not be very elegant, but it used to work.

**Comment by EFLS** [comment]
> so I am unsure what effect the command emacs (“25.1”) would have.

That's just to indicate the minimum required version.

> :commit "271ea573b0a4f265d16108db2ec7c928f3e9aa31"

Apparently, you've got a version from back in April... Not sure how that happened, but there is something wrong with your installation.

Can you find out which `zetteldeft.el` is loaded? I mean, what its location is? Is it the one in your MELPA folder? My guess is that there are two versions in your load path, and the older is loaded. I'm no emacs expert myself, so not sure what the cause is.

**Comment by Phil2740** [comment]
I’m no Emacs expert either.

I did a find on my system, it found only 1 zetteldeft.el, so I don’t think there are multiple instances on my system.

So what I did was downloaded zetteldeft.el from your Github site.

Now everything seems to work OK except when I tried a zetteldeft-follow link I get this error:

Debugger entered--Lisp error: (void-function zetteldeft-link-regex)
  zetteldeft-link-regex()

I see that, at line 86 in your code you have the line:
 
(link-id-re (zetteldeft-link-regex))

Elsewhere in your code the function is zetteldeft--link-regex, 2 hyphens between zetteldeft and link. 

Is this likely to cause a problem?

**Comment by EFLS** [comment]
Glad you've got it sorted now.

> Debugger entered--Lisp error: (void-function zetteldeft-link-regex)

Whoops, I thought I had tested that new code. Should indeed be the `--` variant. Fixed now.

**Comment by Phil2740** [comment]
Thanks for your time sorting this out.

I’ve got another comment to make.

I’m not sure about the avy package. I added this code to my init.el: 

(declare-function avy-jump "avy")
(unless (fboundp 'avy-jump)
  (display-warning 'zetteldeft
    "Function `avy-jump' not available. Please update `avy'”))

I upgraded avy to 20200624.1148, but I still get the prompt to upgrade. avy-related commands seem to do what they are supposed to do. Do I still need that function in my init file?
